src/common/config.cc

```diff
@@ -1218,7 +1220,7 @@ Option::value_t md_config_t::_expand_meta(
       } else if (var == "pid") {
 	out += stringify(getpid());
         if (o) {
-          may_reexpand_meta[o->name] = *str;
+          may_reexpand_meta.push_back(o->name);
```

Comment:
If I override my config (at some level) to `admin socket = /tmp/$pid.asok`, is that original template kept (at its level) and only the expanded version is stored at `CONF_OVERRIDE`?

Reply:
```
enum {
  CONF_DEFAULT,                                                                                                                           
  CONF_MON,
  CONF_FILE,
  CONF_ENV,
  CONF_CMDLINE,
  CONF_OVERRIDE,
  CONF_FINAL
};
```

No, all the original template values of different level will be kept, and the `CONF_OVERRIDE` level will be used in case in runtime setting. Every time when the `_get_val()` is called, it will choose the value in highest priority level if the level is not specified, expand and then return it, at the same time if the chosed value do not contain the `$pid`, it won't touch the  `may_reexpand_meta` map, that means the `may_reexpand_meta` may continue keep the original template value in low level.

Reply:
The `admin socket` is startup type option and there is not observer call back registered.

Reply:
The `may_reexpand_meta` map should just record the options which need to reexpand or to be replaced by a higher priority value if there has.

Reply:
OK, `admin socket` might have been a bad example. However, when I played tested this, it correctly kept my config overrides but it no longer replaced `$pid` with the new daemon pid:

```
$ rbd-mirror --admin-socket ~/tmp.rbd_mirror/test.\$pid.asok --log-file ~/tmp.rbd_mirror/test.\$pid.log --pid-file ~/tmp.rbd_mirror/test.\$pid.pid
$ ll ~/tmp.rbd_mirror/test*
srwxrwxr-x. 1 jdillaman jdillaman    0 Nov 24 12:59 /home/jdillaman/tmp.rbd_mirror/test.959387.asok
-rw-r--r--. 1 jdillaman jdillaman 5348 Nov 24 12:59 /home/jdillaman/tmp.rbd_mirror/test.959387.log
-rw-r--r--. 1 jdillaman jdillaman    7 Nov 24 12:59 /home/jdillaman/tmp.rbd_mirror/test.959387.pid
[jdillaman@cube-1 build]$ ps -ef | grep rbd-mirror
jdillam+  959396       1  0 12:58 ?        00:00:00 rbd-mirror --admin-socket /home/jdillaman/tmp.rbd_mirror/test.$pid.asok --log-file /home/jdillaman/tmp.rbd_mirror/test.$pid.log --pid-file /home/jdillaman/tmp.rbd_mirror/test.$pid.pid --log-to-stderr=true --debug-asok=20 --debug-log=20
```

Reply:
Actually it's working.

My `ceph.conf` for the `admin socket` is:

 ```
34 [client]
 35         keyring = /data/ceph/build/keyring
 36         log file = /data/ceph/build/out/$name.$pid.log
 37         admin socket = /tmp/ceph-asok.2Ze9vx/$name.$pid.asok
```
Run without specifying the `admin socket`, etc:
```
# ./bin/rbd-mirror -d
[root@lxbceph1 src]# ps -aux |grep mirror
root      945472  1.5  0.2 446916 69440 pts/2    Sl+  10:40   0:00 ./bin/rbd-mirror -d

# ls /tmp/ceph-asok.2Ze9vx/
client.admin.945472.asok
```

With your command:

```
# ./bin/rbd-mirror --admin-socket /tmp/tmp.rbd_mirror/test.\$pid.asok --log-file /tmp/tmp.rbd_mirror/test.\$pid.log --pid-file /tmp/tmp.rbd_mirror/test.\$pid.pid

# ps -aux |grep mirror
root      945664  0.8  0.1 446912 39712 ?        Ssl  10:53   0:00 ./bin/rbd-mirror --admin-socket /tmp/tmp.rbd_mirror/test.$pid.asok --log-file /tmp/tmp.rbd_mirror/test.$pid.log --pid-file /tmp/tmp.rbd_mirror/test.$pid.pid

# ls /tmp/tmp.rbd_mirror/
test.945655.log  test.945664.asok  test.945664.log  test.945664.pid
```

Reply:
```
...
> [jdillaman@cube-1 build]$ ps -ef | grep rbd-mirror
> jdillam+  959396       1  0 12:58 ?        00:00:00 rbd-mirror --admin-socket /home/jdillaman/tmp.rbd_mirror/test.$pid.asok --log-file /home/jdillaman/tmp.rbd_mirror/test.$pid.log --pid-file /home/jdillaman/tmp.rbd_mirror/test.$pid.pid --log-to-stderr=true --debug-asok=20 --debug-log=20
 ```
Ceph code won't replace anything of the process's commandline for the ps output.

Reply:
> Ceph code won't replace anything in the process's commandline for the ps output.

Of course, I was not suggesting that it would. I was showing that the process ID was `959396` but the pid and log were still using the original pid of `959387`

Reply:
... I rebuilt my test branch w/ this PR again and it seems to be working this time. :man_shrugging:

Reply:
> > Ceph code won't replace anything in the process's commandline for the ps output.
> 
> Of course, I was not suggesting that it would. I was showing that the process ID was `959396` but the pid and log were still using the original pid of `959387`

Sorry, misreaded it.